pkg/sql/schemachanger/scplan/internal/opgen/opgen_function.go

```diff
@@ -13,21 +13,32 @@ package opgen
 import (
 	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scop"
 	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scpb"
+	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
 )
 
 func init() {
 	opRegistry.register((*scpb.Function)(nil),
 		toPublic(
 			scpb.Status_ABSENT,
 			equiv(scpb.Status_DROPPED),
+			to(scpb.Status_DESCRIPTOR_ADDING,
+				emit(func(this *scpb.Function) *scop.CreateFunctionDescriptor {
+					return &scop.CreateFunctionDescriptor{
+						Function: *protoutil.Clone(this).(*scpb.Function),
+					}
+				}),
+			),
 			to(scpb.Status_PUBLIC,
```

Comment:
This needs to be `revertible(false)`. Or perhaps the previous transition. See what makes sense by looking into something like
```
BEGIN;
CREATE FUNCTION f ...
CREATE UNIQUE INDEX idx ON t(x);
COMMIT;
```
You want the function to become visible to the rest of the cluster only once the schema change reaches the PostCommitNonRevertable phase.

Reply:
Good call, will adjust and add test case for that.

Reply:
done. though looks like there isn't a way to add such test since we still fall back within explicit transaction?

Reply:
If you set the session variable you can do it. `SET use_declarative_schema_changer = unsafe_always`.

Reply:
that is very true.

Reply:
Added a test for this. Awesome work on the execution-time generalization work. This is exciting.

Reply:
hmm.. Actually I should also have a test with the job paused instead of fail and rollback. But I found that reveal some problems here.
Like..we should also make setting ObjectParent (which add the function to schema) not revertible, so that the function won't be visible to function resolution. But we cannot do that, because it would failed pre-commit phase due to cross-reference validation on function descriptor (like it cannot have invalid parent schema). We definitely need the descriptor to exists so that we can track descriptors in jobs with no problem. I think one comprise I can do is to have the descriptor's state set to `DROP` when it's first created. The default value of `PUBLIC` doesn't make sense in anymore.

Reply:
Does the existing data-driven framework not implement this? Like if you add an `end_to_end` test, I think it can have multiple statements. Also, if you do that, I'm pretty certain that we'll generate a `pause` case where it'll pause and resume between each of them.

Maybe one thing we need to do is find a way to talk about operations inside the same transaction between the statements. @fqazi do you have thoughts?

Reply:
thanks, I added those auto generated test including the explains(_verbose) tests for multi-statements. It seems good. The explain on the rollback stages are awesome that it shows rollback on descriptors affected by the multiple statements.
I also set the initial state of the descriptor as `ADD`. I think from now being, in declarative schema changer, we need to do the same when creating other descriptors, so I refactor the `newAddingTableError` thing to be not just for table. This solves the cross-reference validation errors I saw earlier.